README.md
# pycleverclicker
click and keypress record and repeat.

### dependancies ###

* pynput - record and automate mouse and keyboard
* Other community or team contact

CHANGELOG.md
### [unreleased]
## Added
- this file - CHANGELOG.md
- myinputdata.py class to (de) serialize each input line
- Ctrl+Shift+R stops recording by raising event!
- Run prints actions to console
- run options! Edit -> Options and prompted on open
- commit messaged add #xx to link to isssue and prefix w/ 'close' to close - merge to master?!
- export!
- import!
- run!
## Changed
- refactored following pylint
- pymybase located in packages module
- pymybase only included needed files (not entire repo!)
- refactored app components and follow cookiecutter
- press and time in script precision 4 decimals
- refactored - removed "input" from class names
## Bug
- MySettingsDialog not capture all (can still drop root and get's shortcuts???)
## Fixed
- no keyboard long hold due to keyboard - increase C/S override caused issues
- pressing enter on settings dialog at start of recording isn't the first action!
- If recorder is stopped by click, that last click is not added to the list!
## Removed

mydata.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
@author: Elias Wood <owns13927@yahoo.com>
Created on 2017-03-26
store,export,import,iter through scripts
"""

import json
from packages.pymybase.myloggingbase import MyLoggingBase

class MyData(list,MyLoggingBase):
    """class to hold input value - to be passed around"""

    def __init__(self, *args,**keys):
        MyLoggingBase.__init__(self,*args,**keys)

        list.__init__(self,*args,**keys)

    def to_json(self):
        """serialize"""
        for i in self:
            yield json.dumps(i, sort_keys=True, separators=(',', ':'))

    def from_json(self, s):
        """de-serialize"""
        raise NotImplementedError()

    def __str__(self, *args, **kwargs):
        """see to_str"""
        return MyLoggingBase.__str__(self, *args, **kwargs)
    

myplayer.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
@author: Elias Wood <owns13927@yahoo.com>
Created on 2017-03-26
<desc here>
"""

import time
import random
from threading import Thread, Event
from pynput import mouse, keyboard

from packages.pymybase.myloggingbase import MyLoggingBase

class MyPlayer(Thread,MyLoggingBase):
    """play recording mouse and keyboard inputs"""
    
    PRECISION = 4
    UPDATE_EVERY = 1
    __evt = None
    __root = None
    __queue = None
    __actions = None
    __settings = None

    def __init__(self,root, actions_queue, actions, settings, *args,**keys):
        MyLoggingBase.__init__(self,*args,**keys)
        Thread.__init__(self,*args,**keys)

        self.__root = root
        self.__evt = Event()
        self.__queue = actions_queue
        self.__actions = actions
        self.__settings = settings
    
    #===========================================================================
    # STOP
    #===========================================================================
    def stop(self):
        """ start stopping ..."""
        self.__evt.set()
        self.logger.info('told to stop ...')

    #===========================================================================
    # RUN
    #===========================================================================
    def run(self):
        """ do work here """
        self.logger.info('started')
        
        #=======================================================================
        # start controllers
        #=======================================================================
        m = mouse.Controller()
        k = keyboard.Controller()
        self.logger.debug('keyboard and mouse controller set')
        
        #=======================================================================
        # do automation
        #=======================================================================
        # wait delay
        repeat =  self.__settings['repeat limit']
        start_delay = self.__settings['start delay']
        start_var = self.__settings['start variance']
        action_var = self.__settings['action variance']
        
        # keep going
        msg_prefix = '{0:03}/{1:03}' if repeat else '{0:03}'
        
        run_count = 1
        while not self.__evt.is_set():
            # start delay?
            t = start_delay + random.uniform(-start_var, start_var)
            if t > 0:
                self.logger.debug('%03d sleeping start delay, %.3f', run_count, t)
                # add to queue so app can log to console
                self.__queue.put_nowait(dict(
                    run = msg_prefix.format(run_count,run_count),
                    msg = ' starting in {0:02} ...',
                    time = time.time() + t))
                if self.__evt.wait(t): break
            else:
                self.logger.info('%03d sleeping start delay, %.3f', 
                                 run_count, start_delay)
            
            # told to stop?
            if self.__evt.is_set(): break
            
            # go though each action
            for action in self.__actions:
                t = action['time'] + random.uniform(-action_var, action_var)
                
                # add to queue so app can log to console
                self.__queue.put_nowait(dict(
                    run = msg_prefix.format(run_count,run_count),
                    msg = '"'+action['name']+'" in {0:02} ...',
                    time = time.time() + t))
                
                # wait if needed
                if t > 0 and self.__evt.wait(t): break
                
                # do action
                action_name = action['name']
                if action['type'] == 'keyboard':
                    t = action['pressed']
                    self.logger.debug('%.3f %s released after %.1f',
                                     action['time'],action_name,t)
                    k.press(action_name if len(action_name)==1 else keyboard.Key[action_name]) # press
                    if t > 0 and self.__evt.wait(t): break # hold press
                    k.release(action_name if len(action_name)==1 else keyboard.Key[action_name]) # release
                    
                elif action['type'] == 'mouse':
                    self.logger.debug('%.3f %s click at %s',
                                     action['time'],action_name,action['pos'])
                    m.position = action['pos'] # move
                    m.click(mouse.Button[action_name]) # click
                    
                else:
                    self.logger.error('unable to perform action type "%s"',action['type'])
                    
                # told to stop?
                if self.__evt.is_set(): break
            
            # continue?
            if run_count == repeat: break
            run_count += 1
        
        # all done
        self.logger.info('stopped')







myrecorder.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
@author: Elias Wood <owns13927@yahoo.com>
Created on 2017-03-26
<desc here>
"""

import time
from threading import Thread, Lock
from pynput import mouse, keyboard

from packages.pymybase.myloggingbase import MyLoggingBase

class MyRecorder(Thread,MyLoggingBase):
    """handle recording mouse and keyboard input"""
    
    PRECISION = 4
    __continue_running = True
    __queue = None
    __root = None
    __time_lock = None
    __key_lock = None
    __pressed_keys = None
    __last_time = None
    log_clicks = True
    log_keys = True
    ignore_key_press = False # track time a key is pressed
    
    __is_shift = None
    
    IGNORE_KEYS = (keyboard.Key.shift,keyboard.Key.shift_l,keyboard.Key.shift_r,
                   keyboard.Key.ctrl,keyboard.Key.ctrl_l,keyboard.Key.ctrl_r,
                   keyboard.Key.alt,keyboard.Key.alt_gr,keyboard.Key.alt_l,keyboard.Key.alt_r,
                   keyboard.Key.cmd,keyboard.Key.cmd_l,keyboard.Key.cmd_r)

    def __init__(self,root, actions_queue,*args,**keys):
        MyLoggingBase.__init__(self,*args,**keys)
        Thread.__init__(self,*args,**keys)

        self.__root = root
        self.__queue = actions_queue
        self.__time_lock = Lock()
        self.__key_lock = Lock()

    #===========================================================================
    # on_click
    #===========================================================================
    def on_click(self,x_pos,y_pos,button,pressed):
        """handle mouse click events"""
        if not self.__continue_running: return False # stop
        if self.log_clicks: self.logger.debug('mouse %s click %s at [%s,%s]',
                button.name,'pressed' if pressed else 'released',x_pos,y_pos)
        if not pressed: # only trigger releases (up)
            with self.__time_lock: #pylint: disable=not-context-manager
                new_time = time.time()
                if not self.__last_time: self.__last_time = new_time
                self.__queue.put_nowait(dict(type='mouse',name=button.name,
                                             pos=(x_pos,y_pos),
                                             time=round(new_time-self.__last_time,self.PRECISION)))
                #self.__queue.put_nowait('{0!s} click at {1!s}'.format(button.name,(x,y)))
                self.__last_time = new_time

    #===========================================================================
    # on_press
    #===========================================================================
    def on_press(self,key):
        """handle key press"""
        if not self.__continue_running: return False # stop listener
        
        if key == keyboard.Key.shift: self.__is_shift = True
        
        if self.ignore_key_press: return None # do nothing

        # valid key (we want)?
        if key in keyboard.Key:
            if key in self.IGNORE_KEYS: return None # stop
        else:
            k = key.char
            if not k.isalnum(): # or k != k.lower():
                if self.__is_shift and key == keyboard.KeyCode(char='\x12'): #ctrl+r
                    self.logger.debug('sending <<stop-recording>> event ...')
                    self.__root.event_generate('<<stop-recording>>', when="tail")
                    return False # stop listener
                self.logger.warning('invalid key %r',key)
                return None # stop


        # track time when pressed
        name =  key.char if hasattr(key,'char') else key.name
        with self.__key_lock: #pylint: disable=not-context-manager
            cur_time = time.time()
            set_time = self.__pressed_keys.setdefault(name,cur_time)
        
        # if we're logging keys and this is the first time it's been pressed, log it!
        if self.log_keys and cur_time == set_time:
            self.logger.debug('key %s pressed',key.name if hasattr(key,'name') else key.char)

    #===========================================================================
    # on_release
    #===========================================================================
    def on_release(self,key):
        """handle key release"""
        if not self.__continue_running: return False # stop
        
        if key == keyboard.Key.shift: self.__is_shift = False
        
        # valid key (we want)?
        if key in keyboard.Key:
            if key in self.IGNORE_KEYS: return None # stop
        else:
            k = key.char
            if not k.isalnum(): # or k != k.lower():
                self.logger.warning('invalid key %r',key)
                return None # stop

        if self.log_keys:
            self.logger.debug('key %s released',key.name if hasattr(key,'name') else key.char)

        with self.__time_lock: #pylint: disable=not-context-manager
            #  name (unique name) of key pressed
            name=key.char if hasattr(key,'char') else key.name
            # time sense last action?
            new_time = time.time()
            if not self.__last_time: self.__last_time = new_time
            # how long was it pressed?
            if self.ignore_key_press: pressed = 0
            else:
                with self.__key_lock: #pylint: disable=not-context-manager
                    pressed = new_time - self.__pressed_keys.pop(name,new_time)
            # add to queue
            self.__queue.put_nowait(dict(type='keyboard',name=name,
                                         time=round(new_time-self.__last_time,self.PRECISION),
                                         pressed=round(pressed,self.PRECISION)))
            # store time for next action
            self.__last_time = new_time

    #===========================================================================
    # STOP
    #===========================================================================
    def stop(self):
        """ start stopping ..."""
        self.logger.info('told to stop ...')
        self.__continue_running = False

    #===========================================================================
    # RUN
    #===========================================================================
    def run(self):
        """ do work here """
        self.logger.info('started')
        self.__pressed_keys = dict()
        self.__last_time = None

        #=======================================================================
        # # start listeners
        #=======================================================================
        list_m = mouse.Listener(on_click=self.on_click)
        list_m.start()
        list_m.wait()
        list_k = keyboard.Listener(on_release=self.on_release,
                                   on_press=self.on_press)
        list_k.start()
        list_k.wait()

        #=======================================================================
        # wait to stop
        #=======================================================================
        while self.__continue_running: time.sleep(.5)

        #=======================================================================
        # stop listeners
        #=======================================================================
        # NOTE: ubuntu will wait for input before really stopping ...
        list_m.stop()
        mouse.Controller().move(1,1) # ubuntu will wait for input before really stopping ...
        self.logger.debug('joining mouse listener...')
        list_m.join()
        list_k.stop()
        k = keyboard.Controller() # ubuntu will wait for input before really stopping ...
        k.press(keyboard.Key.shift)
        k.release(keyboard.Key.shift)
        self.logger.debug('joining keyboard listener...')
        list_k.join()

        self.logger.info('stopped')







myinputdata.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
@author: Elias Wood <owns13927@yahoo.com>
Created on 2017-03-26
store,export,import,iter through scripts
"""

import json
from packages.pymybase.myloggingbase import MyLoggingBase

class MyInputData(list,MyLoggingBase):
    """class to hold input value - to be passed around"""

    def __init__(self, *args,**keys):
        MyLoggingBase.__init__(self,*args,**keys)

        list.__init__(self,*args,**keys)

    def to_json(self):
        """serialize"""
        for i in self:
            yield json.dumps(i, sort_keys=True, separators=(',', ':'))

    def from_json(self, s):
        """de-serialize"""
        raise NotImplementedError()

    def __str__(self, *args, **kwargs):
        """see to_str"""
        return MyLoggingBase.__str__(self, *args, **kwargs)
    

mysettingsdialog.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
@author: Elias Wood <owns13927@yahoo.com>
Created on 2018-02-10
store,export,import,iter through scripts
"""

from packages.pymybase.myloggingbase import MyLoggingBase
import tkinter as tk
import tkinter.ttk as ttk
#from idlelib.tooltip import ToolTip
from tkinter.simpledialog import Dialog

class MySettingsDialog(Dialog,MyLoggingBase):
    
    result = None
    __inputs = None
    __entries = None
    
    def __init__(self,parent,title=None,
                 inputs=None,
                 name=None):
        MyLoggingBase.__init__(self,name=name)
        self.__inputs = inputs.copy()
        
        Dialog.__init__(self,parent=parent,title=title)
        
        
    def body(self, master):
        '''create dialog body.

        return widget that should have initial focus.
        This method should be overridden, and is called
        by the __init__ method.
        '''
        Dialog.body(self, master)
        
        # set icon
        try: self.wm_iconbitmap(self.get_resource_fd('python-icon.ico'))
        except tk.TclError: pass # in ubuntu: _tkinter.TclError: bitmap /.../... not defined
        
        # add inputs
        self.__entries = dict()
        first_key = None
        r = 0
        for k,v in self.__inputs.items():
            # get first entry added so we can set that to have focus
            if first_key is None: first_key = k
            
            # create label
            ttk.Label(master, text=k+':',justify=tk.LEFT).grid(row=r,column=0,sticky=tk.W)
            
            # create entry
            e = ttk.Entry(master,width=9)
            e.grid(row=r,column=1)
            # set default input if any
            if v is not None: e.insert(0, v)
            # add to list
            self.__entries[k] = e
            
            # next row!
            r += 1
        
        # set focus to first entry
        if self.__entries: return self.__entries[first_key]
    
    def buttonbox(self):
        '''add standard button box.

        override if you do not want the standard buttons
        '''
        #Dialog.buttonbox(self)
        
        # override to use ttk buttons!
        box = ttk.Frame(self)

        w = ttk.Button(box, text="OK", width=10, command=self.ok, default=tk.ACTIVE)
        w.pack(side=tk.LEFT, padx=5, pady=5)
        w = ttk.Button(box, text="Cancel", width=10, command=self.cancel)
        w.pack(side=tk.LEFT, padx=5, pady=5)
        
        self.bind("<KeyRelease-Return>", self.ok) #<Return>
        self.bind("<Escape>", self.cancel)

        box.pack()
     
    def validate(self):
        '''validate the data

        This method is called automatically to validate the data before the
        dialog is destroyed. By default, it always validates OK.
        '''
        if Dialog.validate(self) == 0: return 0
        
        for k,entry in self.__entries.items():
            s = entry.get()
            
            # validate if needed
            if self.__inputs[k] is not None and self.__inputs[k] != s:
                # validate data type
                t = type(self.__inputs[k])
                try:
                    v = t(s)
                except ValueError as e:
                    # failed!
                    tk.messagebox.showwarning(
                        e.__class__.__name__,
                        "'{0}' must be of type '{1}'!\nPlease try again".format(k,t.__name__),
                        parent = self
                    )
                    return 0
                else:
                    # good to go, override default with value
                    self.__inputs[k] = v
        
        # passed all validation! - 'apply' to result
        self.result = self.__inputs
        return 1
        
    def apply(self):
        '''process the data

        This method is called automatically to process the data, *after*
        the dialog is destroyed. By default, it does nothing.
        '''
        Dialog.apply(self)
        

myinputrecorder.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
@author: Elias Wood <owns13927@yahoo.com>
Created on 2017-03-26
<desc here>
"""

import time
from threading import Thread, Lock
from pynput import mouse, keyboard

from packages.pymybase.myloggingbase import MyLoggingBase

class MyInputRecorder(Thread,MyLoggingBase):
    """handle recording mouse and keyboard input"""
    
    PRECISION = 4
    __continue_running = True
    __queue = None
    __root = None
    __time_lock = None
    __key_lock = None
    __pressed_keys = None
    __last_time = None
    log_clicks = True
    log_keys = True
    ignore_key_press = False # track time a key is pressed
    
    __is_shift = None
    
    IGNORE_KEYS = (keyboard.Key.shift,keyboard.Key.shift_l,keyboard.Key.shift_r,
                   keyboard.Key.ctrl,keyboard.Key.ctrl_l,keyboard.Key.ctrl_r,
                   keyboard.Key.alt,keyboard.Key.alt_gr,keyboard.Key.alt_l,keyboard.Key.alt_r,
                   keyboard.Key.cmd,keyboard.Key.cmd_l,keyboard.Key.cmd_r)

    def __init__(self,root, actions_queue,*args,**keys):
        MyLoggingBase.__init__(self,*args,**keys)
        Thread.__init__(self,*args,**keys)

        self.__root = root
        self.__queue = actions_queue
        self.__time_lock = Lock()
        self.__key_lock = Lock()

    #===========================================================================
    # on_click
    #===========================================================================
    def on_click(self,x_pos,y_pos,button,pressed):
        """handle mouse click events"""
        if not self.__continue_running: return False # stop
        if self.log_clicks: self.logger.debug('mouse %s click %s at [%s,%s]',
                button.name,'pressed' if pressed else 'released',x_pos,y_pos)
        if not pressed: # only trigger releases (up)
            with self.__time_lock: #pylint: disable=not-context-manager
                new_time = time.time()
                if not self.__last_time: self.__last_time = new_time
                self.__queue.put_nowait(dict(type='mouse',name=button.name,
                                             pos=(x_pos,y_pos),
                                             time=round(new_time-self.__last_time,self.PRECISION)))
                #self.__queue.put_nowait('{0!s} click at {1!s}'.format(button.name,(x,y)))
                self.__last_time = new_time

    #===========================================================================
    # on_press
    #===========================================================================
    def on_press(self,key):
        """handle key press"""
        if not self.__continue_running: return False # stop listener
        
        if key == keyboard.Key.shift: self.__is_shift = True
        
        if self.ignore_key_press: return None # do nothing

        # valid key (we want)?
        if key in keyboard.Key:
            if key in self.IGNORE_KEYS: return None # stop
        else:
            k = key.char
            if not k.isalnum(): # or k != k.lower():
                if self.__is_shift and key == keyboard.KeyCode(char='\x12'): #ctrl+r
                    self.logger.debug('sending <<stop-recording>> event ...')
                    self.__root.event_generate('<<stop-recording>>', when="tail")
                    return False # stop listener
                self.logger.warning('invalid key %r',key)
                return None # stop


        # track time when pressed
        name =  key.char if hasattr(key,'char') else key.name
        with self.__key_lock: #pylint: disable=not-context-manager
            cur_time = time.time()
            set_time = self.__pressed_keys.setdefault(name,cur_time)
        
        # if we're logging keys and this is the first time it's been pressed, log it!
        if self.log_keys and cur_time == set_time:
            self.logger.debug('key %s pressed',key.name if hasattr(key,'name') else key.char)

    #===========================================================================
    # on_release
    #===========================================================================
    def on_release(self,key):
        """handle key release"""
        if not self.__continue_running: return False # stop
        
        if key == keyboard.Key.shift: self.__is_shift = False
        
        # valid key (we want)?
        if key in keyboard.Key:
            if key in self.IGNORE_KEYS: return None # stop
        else:
            k = key.char
            if not k.isalnum(): # or k != k.lower():
                self.logger.warning('invalid key %r',key)
                return None # stop

        if self.log_keys:
            self.logger.debug('key %s released',key.name if hasattr(key,'name') else key.char)

        with self.__time_lock: #pylint: disable=not-context-manager
            #  name (unique name) of key pressed
            name=key.char if hasattr(key,'char') else key.name
            # time sense last action?
            new_time = time.time()
            if not self.__last_time: self.__last_time = new_time
            # how long was it pressed?
            if self.ignore_key_press: pressed = 0
            else:
                with self.__key_lock: #pylint: disable=not-context-manager
                    pressed = new_time - self.__pressed_keys.pop(name,new_time)
            # add to queue
            self.__queue.put_nowait(dict(type='keyboard',name=name,
                                         time=round(new_time-self.__last_time,self.PRECISION),
                                         pressed=round(pressed,self.PRECISION)))
            # store time for next action
            self.__last_time = new_time

    #===========================================================================
    # STOP
    #===========================================================================
    def stop(self):
        """ start stopping ..."""
        self.logger.info('told to stop ...')
        self.__continue_running = False

    #===========================================================================
    # RUN
    #===========================================================================
    def run(self):
        """ do work here """
        self.logger.info('started')
        self.__pressed_keys = dict()
        self.__last_time = None

        #=======================================================================
        # # start listeners
        #=======================================================================
        list_m = mouse.Listener(on_click=self.on_click)
        list_m.start()
        list_m.wait()
        list_k = keyboard.Listener(on_release=self.on_release,
                                   on_press=self.on_press)
        list_k.start()
        list_k.wait()

        #=======================================================================
        # wait to stop
        #=======================================================================
        while self.__continue_running: time.sleep(.5)

        #=======================================================================
        # stop listeners
        #=======================================================================
        # NOTE: ubuntu will wait for input before really stopping ...
        list_m.stop()
        mouse.Controller().move(1,1) # ubuntu will wait for input before really stopping ...
        self.logger.debug('joining mouse listener...')
        list_m.join()
        list_k.stop()
        k = keyboard.Controller() # ubuntu will wait for input before really stopping ...
        k.press(keyboard.Key.shift)
        k.release(keyboard.Key.shift)
        self.logger.debug('joining keyboard listener...')
        list_k.join()

        self.logger.info('stopped')





